File: 05g-kafka-cloudevents-avro-with-springboot/src/python-cloudevents-kafka-producer-kafka-native/cloud_event_producer.py
import uuid
from datetime import datetime, timezone

from confluent_kafka import Producer
from confluent_kafka.schema_registry import SchemaRegistryClient
from confluent_kafka.schema_registry.avro import AvroSerializer
from confluent_kafka.serialization import SerializationContext, MessageField

_CE_SPEC_VERSION = "ce_specversion"
_CE_ID           = "ce_id"
_CE_TYPE         = "ce_type"
_CE_SOURCE       = "ce_source"
_CE_TIME         = "ce_time"
_CE_DATA_SCHEMA  = "ce_dataschema"
_CONTENT_TYPE    = "content-type"


class CloudEventProducer:

    def __init__(self, bootstrap_servers: str, schema_registry_url: str, source: str):
        self._source = source
        self._schema_registry_url = schema_registry_url
        self._schema_registry_client = SchemaRegistryClient({"url": schema_registry_url})
        self._producer = Producer({"bootstrap.servers": bootstrap_servers})
        self._serializer_cache: dict[str, AvroSerializer] = {}

    def publish(
        self,
        topic: str,
        key: str,
        payload: dict,
        schema_str: str,
        event_type: str,
        subject_name: str | None = None,
    ) -> None:
        serializer = self._get_serializer(schema_str)
        value_bytes = serializer(payload, SerializationContext(topic, MessageField.VALUE))

        headers = self._build_ce_headers(event_type, subject_name)

        self._producer.produce(
            topic=topic,
            key=key,
            value=value_bytes,
            headers=headers,
            on_delivery=_delivery_callback,
        )
        self._producer.poll(0)

    def flush(self) -> None:
        self._producer.flush()

    def _get_serializer(self, schema_str: str) -> AvroSerializer:
        if schema_str not in self._serializer_cache:
            self._serializer_cache[schema_str] = AvroSerializer(
                self._schema_registry_client, schema_str
            )
        return self._serializer_cache[schema_str]

    def _build_ce_headers(
        self, event_type: str, subject_name: str | None
    ) -> list[tuple[str, str]]:
        headers = [
            (_CE_SPEC_VERSION, "1.0"),
            (_CE_ID,           str(uuid.uuid4())),
            (_CE_TYPE,         event_type),
            (_CE_SOURCE,       self._source),
            (_CE_TIME,         datetime.now(timezone.utc).isoformat()),
            (_CONTENT_TYPE,    "avro/binary"),
        ]
        if subject_name:
            data_schema = (
                f"{self._schema_registry_url}/subjects/{subject_name}/versions/latest"
            )
            headers.append((_CE_DATA_SCHEMA, data_schema))
        return headers


def _delivery_callback(err, msg) -> None:
    if err:
        print(f"ERROR: delivery failed [key={msg.key()}]: {err}")
    else:
        print(
            f"Published CE [key={msg.key()}, topic={msg.topic()}, "
            f"partition={msg.partition()}, offset={msg.offset()}]"
        )

File: 05g-kafka-cloudevents-avro-with-springboot/src/spring-boot-cloudevents-kafka-producer-kafka-native/src/main/java/com/trivadis/kafkaws/events/CloudEventProducer.java
package com.trivadis.kafkaws.events;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.avro.specific.SpecificRecord;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.header.internals.RecordHeaders;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@Slf4j
@Component
@RequiredArgsConstructor
public class CloudEventProducer {

    private final KafkaTemplate<String, Object> kafkaTemplate;

    @Value("${cloudevents.source}")
    private String defaultSource;

    @Value("${cloudevents.schema-registry-base-url}")
    private String schemaRegistryBaseUrl;

    /**
     * Publish an Avro record as a CloudEvent in binary mode.
     *
     * @param topic       target Kafka topic
     * @param key         Kafka message key (typically aggregate ID)
     * @param payload     Avro SpecificRecord (serialized by KafkaAvroSerializer)
     * @param eventType   CE type, e.g. "com.example.orders.OrderCreated"
     * @param subjectName Schema Registry subject for ce_dataschema header
     */
    public <T extends SpecificRecord> CompletableFuture<SendResult<String, Object>> publish(
            String topic,
            String key,
            T payload,
            String eventType,
            String subjectName) {

        RecordHeaders headers = buildCeHeaders(eventType, subjectName);

        ProducerRecord<String, Object> record =
            new ProducerRecord<>(topic, null, key, payload, headers);

        return kafkaTemplate.send(record)
            .whenComplete((result, ex) -> {
                if (ex != null) {
                    log.error("Failed to publish CE [type={}, key={}]: {}",
                        eventType, key, ex.getMessage(), ex);
                } else {
                    log.info("Published CE [type={}, key={}, partition={}, offset={}]",
                        eventType, key,
                        result.getRecordMetadata().partition(),
                        result.getRecordMetadata().offset());
                }
            });
    }

    // Overload without subjectName (omits ce_dataschema header)
    public <T extends SpecificRecord> CompletableFuture<SendResult<String, Object>> publish(
            String topic, String key, T payload, String eventType) {
        return publish(topic, key, payload, eventType, null);
    }

    // ---------------------------------------------------------------

    private RecordHeaders buildCeHeaders(String eventType, String subjectName) {
        RecordHeaders headers = new RecordHeaders();
        headers.add(CloudEventHeaders.SPEC_VERSION, utf8(CloudEventHeaders.SPEC_VERSION_1_0));
        headers.add(CloudEventHeaders.ID,           utf8(UUID.randomUUID().toString()));
        headers.add(CloudEventHeaders.TYPE,         utf8(eventType));
        headers.add(CloudEventHeaders.SOURCE,       utf8(defaultSource));
        headers.add(CloudEventHeaders.TIME,         utf8(Instant.now().toString()));
        headers.add(CloudEventHeaders.CONTENT_TYPE, utf8(CloudEventHeaders.AVRO_CONTENT_TYPE));

        if (subjectName != null) {
            String dataSchema = schemaRegistryBaseUrl
                + "/subjects/" + subjectName + "/versions/latest";
            headers.add(CloudEventHeaders.DATA_SCHEMA, utf8(dataSchema));
        }
        return headers;
    }

    private static byte[] utf8(String value) {
        return value.getBytes(StandardCharsets.UTF_8);
    }
}